build.ps1

```diff
@@ -1,66 +1,140 @@
+    [Alias('u')]
+    [switch]$Update,
+    [Parameter(ValueFromRemainingArguments = $true)]
+    [string[]]$MSBuildArgs
+)
 
-    New-Item -Path "$buildFolder" -Type directory | Out-Null
-    copy-item "$tempFolder\**\build\*" $buildFolder -Recurse
+Set-StrictMode -Version 2
+$ErrorActionPreference = 'Stop'
 
-    # Cleanup
-    if (Test-Path $tempFolder) {
-        Remove-Item -Recurse -Force $tempFolder
+#
+# Functions
+#
+
+function Get-KoreBuild {
+
+    $lockFile = Join-Path $Path 'korebuild-lock.txt'
+
+    if (!(Test-Path $lockFile) -or $Update) {
+        Get-RemoteFile "$ToolsSource/korebuild/channels/$Channel/latest.txt" $lockFile
     }
+
+    $version = Get-Content $lockFile -Tail 1
+    $korebuildPath = Join-Paths $DotNetHome ('buildtools', 'korebuild', $version)
+
+    if (!(Test-Path $korebuildPath)) {
+        Write-Host "Downloading KoreBuild $version"
+        New-Item -ItemType Directory -Path $korebuildPath | Out-Null
+        $remotePath = "$ToolsSource/korebuild/artifacts/$version/korebuild.$version.zip"
+
+        try {
+            if (Get-Command -Name 'Expand-Archive' -ErrorAction Ignore) {
+                # Use built-in commands where possible as they are cross-plat compatible
+                $tmpfile = "$(New-TemporaryFile).zip"
+                Get-RemoteFile $remotePath $tmpfile
+                Expand-Archive -Path $tmpfile -DestinationPath $korebuildPath
+            } else {
+                # Fallback to old approach for old installations of PowerShell
+                $tmpfile = Join-Path $env:TEMP "KoreBuild-$([guid]::NewGuid()).zip"
+                Get-RemoteFile $remotePath $tmpfile
+                Add-Type -AssemblyName System.IO.Compression.FileSystem
+                [System.IO.Compression.ZipFile]::ExtractToDirectory($tmpfile, $korebuildPath)
+            }
+        }
+        finally {
+            remove-item $tmpfile -ErrorAction Ignore
+        }
+    }
+
+    return $korebuildPath
 }
 
-&"$buildFile" @args
+function Join-Paths($path, $childPaths) {
+    $childPaths | ForEach-Object { $path = Join-Path $path $_ }
+    return $path
+}
+
+function Get-RemoteFile($RemotePath, $LocalPath) {
+    if ($RemotePath -notlike 'http*') {
+        Copy-Item $RemotePath $LocalPath
+        return
+    }
+
+    $retries = 10
+    while ($retries -gt 0) {
+        $retries -= 1
+        try {
+            Invoke-WebRequest -UseBasicParsing -Uri $RemotePath -OutFile $LocalPath
+            break
+        }
+        catch {
+            if ($retries -le 0) {
+                Write-Error "Download failed: '$RemotePath'."
+            }
+            else {
+                Write-Verbose "Request failed. $retries retries remaining"
+            }
+        }
+    }
+}
+
+#
+# Main
+#
+
+try {
+    $korebuildPath = Get-KoreBuild
+    Import-Module -Force -Scope Local (Join-Path $korebuildPath 'KoreBuild.psd1')
+    Install-Tools $ToolsSource $DotNetHome
+    Invoke-RepositoryBuild $Path @MSBuildArgs
```

Comment:
This was very deliberate. Let me explain my thinking:
 - make it clearer to those reading the script which steps will happen when build.ps1 executes.
 - `Install-Tools` and `Invoke-RepositoryBuild` should be treated like an "API". Any changes to arguments or behavior is 'breaking' and should not be done lightly. These will rarely if ever change, so our bootstrappers can delegate the majority of stuff to these two functions.
 - I'm totally okay with one extra line in build.ps1.

FYI - I'm also thinking forward to what we discussed in our runtime-engineering meeting last week: https://github.com/aspnet/specs/pull/77. One of the problems we have to day is that our build is a mix of console tools, tasks, and scripts, scattered across nuget packages, source code, and stuff that gets downloaded on the fly. We want KoreBuild capture the "state" of build tools. Decomposing KoreBuild into a module (KoreBuild.ps**m**1) gives us more flexibility to write scripts that re-use tools from KoreBuild. `Korebuild.ps1` was too inflexible. For example, I want to add a function to KoreBuild.psd1 to `Push-Packages`. This could delegate to `nuget.exe`, `dotnet nuget`, or `PackagePublisher.dll`, but it doesn't matter to the caller since all they need is the powershell function.